### ISSUE TITLE
Implement Server Setters; RPC are broken

### ISSUE THREAD

**Issue by ahcenezdh** [citizenfx/fivem#2787]
### What happened?

# Feature Request: Server Setters for Authority and Stability

## Background

For a considerable time, many developers have been requesting Server Setters as an alternative to RPC natives. It's widely acknowledged in the FiveM community that RPC natives are often unstable or broken.

@prikolium-cfx mentioned the reliability issues of RPC natives in [this pull request](https://github.com/citizenfx/fivem/pull/2038).

## Current Issues

1. **Unreliability of RPC Natives:**
   - Blip natives work approximately 20% of the time.
   - `SetPedConfigFlag` and vehicle RPC natives frequently require workarounds. (when they work)
   - Developers often need to implement delays to ensure entity existence before manipulation.

2. **Lack of Server Authority:**
   - Current methods heavily rely on client-side operations, reducing server control.

3. **Code Cleanliness:**
   - Reliance on RPC natives often leads to complex, less maintainable code.

## Proposed Solution

Implement a comprehensive set of Server Setters to complement existing server-side functions.

### Current Server-Side Functions
- `CreateVehicleServerSetter`
- `CreatePed`
- `CreateObjectNoOffset`

While these are beneficial, they are insufficient for a robust server API.

### Requested Server Setters

Please note that this list is not exhaustive and is intended to provide examples of the types of setters that would be useful:

#### Entity Setters
- `SetEntityHeading`
- `SetEntityCoords`
- `SetEntityRotation`
- `SetEntityVelocity`

#### Vehicle Setters
- `SetVehicleNumberPlateText`
- `SetVehicleDoorsLocked`

#### Vehicle Modification Setters
- `GetVehicleMod`
- `GetVehicleModKit`
- `GetVehicleModKitType`
- `GetVehicleModVariation`
- `SetVehicleMod`
- `SetVehicleModKit`

## Benefits

1. **Server Authority:**
   - Direct server control over entity properties and behaviors.

2. **Improved Stability:**
   - Reduction in reliance on unstable RPC natives.

3. **Cleaner Code:**
   - Simplified server-side logic without complex client-side workarounds.

4. **Better Synchronization:**
   - Potential for improved entity sync across clients.

## Use Cases

1. **Vehicle Management:**
   - Server can directly modify vehicle properties (e.g., locking doors, changing plates) without client intervention.

2. **NPC Behavior Control:**
   - Precise control over NPC movements and actions from the server.

3. **Anti-Cheat Measures:**
   - Server can validate and correct entities properties to prevent client-side manipulation.

4. **Manipulating multiple entities :**
   - Server can orchestrate complex scenarios by manipulating multiple entities simultaneously.

### Expected result

Server setters

### Reproduction steps

1. Use RPC natives
2. See rpc natives are broken 
3. Use weird workaround to use them or directly use client side code

### Importancy

There's a workaround

### Area(s)

FiveM, FXServer, OneSync, Natives

### Specific version(s)

Every version

**Comment by DaniGP17** [comment]
I think it is a very good idea, I have had experience when creating scripts that needed to modify the modifications of a vehicle from the server side, I have always had to assign to a client the need to assign these modifications, with this proposal I think it can solve many problems of this style and not depend so much on the clients.

**Comment by Wilfred4** [comment]
I currently have a race server, and that would be a game changer for us, we have so much tricky code and have to do callback from the client, server functions would be awesome

**Comment by Banes-dev** [comment]
Like a lot of developper, i had do a customisation script for vehicle, and the dream of all developper are do this is possibility to put this in server side

**Comment by s1nyx** [comment]
Hi,

Couldn't agree more. It's clearly a **must-have**, not a nice to have.

**Comment by Tygaugh** [comment]
Yes it's a good idea, we need this since a long time ago, like also for the blips ( SetColourBlip, Scale etc...)

**Comment by DevSekai** [comment]
We need this !!

**Comment by pablo-1610** [comment]
Excellent idea, we need more control over clients 👏

**Comment by Mathu-lmn** [comment]
👍🏻 Needed those many times for years !

**Comment by Sayzx** [comment]
👍🏻 Good idea we need this !!!!

**Comment by ekaliclua** [comment]
I completely agree with this, and it’s essential that we implement it to maintain control over the players. 👏

**Comment by matthias18771** [comment]
This would be very sick!

**Comment by yveltdev** [comment]
Nice idea ! <3

**Comment by SertinoxVS** [comment]
It's a realy good idea can had lot of control !

we need this !

**Comment by BuddiesTV** [comment]
This is definitely needed! It would give serverside more control, which is great for a lot of reasons - especially when it comes to dealing with cheaters.

**Comment by ShadowCodding** [comment]
This is a great idea

**Comment by spoty123** [comment]
This feature is one of the most needed and requested features that FiveM can offer.
The current situation is that servers must use client callbacks or statebags (which have their issues), forcing servers to rely on the client to perform simple utilities, such as changing a vehicle's license plate upon spawn. 
Having the ability to perform these kinds of tasks via the server, would simplify the logic of scripts drastically, and would also certainly prevent sync and entity owner issues many developers run into when trying to set features on an entity.

**Comment by OffSey** [comment]
I approve of this wonderful idea! 👍

**Comment by DeVerino-DVR** [comment]
Yes, an excellent idea and nothing but the Simplification of the code I approve

**Comment by MelissssaEL** [comment]
Oh yes, that would be amazing!

**Comment by Dolyyyy** [comment]
This possibility would be incredible. Relying on the client for basic tasks complicates development. Server-side control would simplify scripts (no need to navigate client -> server several times for nothing) and eliminate unnecessary callbacks, for example.

**Comment by BrownyProd** [comment]
Great idea! Implementing **Server Setters** across all entity types, not just vehicles and NPCs, would significantly improve server stability and control. This would reduce reliance on unstable RPC natives, clean up code, and enhance synchronization across clients. Expanding the setter options would also improve server authority over objects, blips, and player properties, making things more reliable and efficient. Looking forward to this!

**Comment by ook3D** [comment]
> Looking forward to this!

you realize this is just an issue right not a pull request?

**Comment by BrownyProd** [comment]
> you realize this is just an issue right not a pull request?

And so I can't look forward to something actually useful being (hope so) implemented?

**Comment by ook3D** [comment]
> > you realize this is just an issue right not a pull request?
> 
> And so I can't look forward to something actually useful being (hope so) implemented?

I mean sure go for it, but you're going to be waiting months

**Comment by ahcenezdh** [comment]
They will be waiting for months

**Comment by Skyness60** [comment]
Good idea ☑️👍🏼

**Comment by myrm1003** [comment]
My server currently uses client-side vehicles, since server vehicles are broken at the moment. The development of a reliable server API for vehicles is needed, along with server setters. I'm looking forward to this.

**Comment by yiruzu** [comment]
This is definitely needed! I hope these get implemented soon!

**Comment by dopyyy** [comment]
Excellent idea, it improves control and security.

**Comment by epyidev** [comment]
Nice idea 👍

**Comment by KillStr3aK** [comment]
> Excellent idea, it improves control and security. I take this example mentioned by @Dolyyyy : "(no need to navigate client -> server several times for nothing) and eliminate unnecessary callbacks, for example."

currently it is implemented like that because entities does not exists on the server side in the way they exists on the clients 

this all sounds good but it wasn't implemented yet for its own reason and not because they just didn't wanted to make peoples life and scripting easier, this will require a lot of planning

**Comment by razakmhd** [comment]
I'm creating mission scripts just like in gta online and i'm waiting for this, on SAMP things were easier :)

**Comment by Shino-World** [comment]
Nice idea bro👍

**Comment by quatropf** [comment]
👍🏻 Good idea we need this !!!!

**Comment by thelindat** [comment]
> 👍🏻 Good idea we need this !!!!

Well thank god you and the 30 other muppsets posted "good idea!!!". Really contributing to the discussion there champ.

**Comment by Pedro-Lucas14** [comment]
Nice idea 👍

**Comment by ook3D** [comment]
> Nice idea 👍

Thanks for the valuable input, we didn't see the first 30 replies from mindless drones flocking here

**Comment by Firou91** [comment]
# Server Setters for Authority and Stability [#2787](https://github.com/citizenfx/fivem/issues/2787) (Response to Feature Request)

I find this request very relevant, as it addresses key issues that developers often encounter due to over-reliance on native RPC. Here are some thoughts and additions on the subject:

## Current problems

### Instability of native RPC
It's true that RPC natives present reliability problems, especially when it comes to server-side entity manipulation. This leads not only to random bugs, but also to inconsistencies in entity synchronization between clients. Many developers circumvent these limitations by using delays and other less-than-optimal methods. It would indeed be beneficial to have reliable server setters to get around these difficulties.

### Lack of Server Authority
Server-side entity management is limited by the current implementation, which leaves a great deal of responsibility to the client. Introducing server-setters would not only improve stability, but also enhance entity security and integrity. This would prevent unwanted client-side manipulation and offer better management of persistent entity states.

## Suggestions for additional features
To complement your list of server setters, it might be interesting to add :

### Setters for Dynamic Entities
- SetPedComponentVariationServerSide**: To manage NPC clothing directly on the server side, which would improve scenarios involving uniforms, for example.
- SetEntityVisibleServerSide**: To manage entity visibility without the need for a client call.

### Improved synchronization
One of the major benefits of server-side setters is improved entity synchronization, particularly in environments with a large number of simultaneous interactions (e.g. RP servers). If the server can handle more entity-related aspects, this would avoid frequent desynchronization, especially on OneSync servers with lots of players.

### Performance optimization
Reducing RPC calls could not only improve stability, but also lighten the load on the client side. By delegating more to the server, this would enable a better distribution of tasks, particularly beneficial for clients with more modest configurations.

## Conclusion

This proposal to add server-side setters would be a major step forward in improving the stability, security and maintainability of FiveM servers. I strongly encourage this initiative and am curious to see how it could be integrated with the framework's current features. 

Thank you for your suggestion, which I hope will be taken on board by the CFX team.

**Comment by Firou91** [comment]
# Server Setters for Authority and Stability (Response to Feature Request)

I find this request very relevant, as it addresses key issues that developers often encounter due to over-reliance on native RPC. Here are some thoughts and additions on the subject:

## Current problems

### Instability of native RPC
It's true that RPC natives present reliability problems, especially when it comes to server-side entity manipulation. This leads not only to random bugs, but also to inconsistencies in entity synchronization between clients. Many developers circumvent these limitations by using delays and other less-than-optimal methods. It would indeed be beneficial to have reliable server setters to get around these difficulties.

### Lack of Server Authority
Server-side entity management is limited by the current implementation, which leaves a great deal of responsibility to the client. Introducing server-setters would not only improve stability, but also enhance entity security and integrity. This would prevent unwanted client-side manipulation and offer better management of persistent entity states.

## Suggestions for additional features
To complement your list of server setters, it might be interesting to add :

### Setters for Dynamic Entities
- SetPedComponentVariationServerSide**: To manage NPC clothing directly on the server side, which would improve scenarios involving uniforms, for example.
- SetEntityVisibleServerSide**: To manage entity visibility without the need for a client call.

### Improved synchronization
One of the major benefits of server-side setters is improved entity synchronization, particularly in environments with a large number of simultaneous interactions (e.g. RP servers). If the server can handle more entity-related aspects, this would avoid frequent desynchronization, especially on OneSync servers with lots of players.

### Performance optimization
Reducing RPC calls could not only improve stability, but also lighten the load on the client side. By delegating more to the server, this would enable a better distribution of tasks, particularly beneficial for clients with more modest configurations.

## Conclusion

This proposal to add server-side setters would be a major step forward in improving the stability, security and maintainability of FiveM servers. I strongly encourage this initiative and am curious to see how it could be integrated with the framework's current features. 

Thank you for your suggestion, which I hope will be taken on board by the CFX team.

**Comment by ook3D** [comment]
> # Server Setters for Authority and Stability (Response to Feature Request)
> I find this request very relevant, as it addresses key issues that developers often encounter due to over-reliance on native RPC. Here are some thoughts and additions on the subject:
> 
> ## Current problems
> ### Instability of native RPC
> It's true that RPC natives present reliability problems, especially when it comes to server-side entity manipulation. This leads not only to random bugs, but also to inconsistencies in entity synchronization between clients. Many developers circumvent these limitations by using delays and other less-than-optimal methods. It would indeed be beneficial to have reliable server setters to get around these difficulties.
> 
> ### Lack of Server Authority
> Server-side entity management is limited by the current implementation, which leaves a great deal of responsibility to the client. Introducing server-setters would not only improve stability, but also enhance entity security and integrity. This would prevent unwanted client-side manipulation and offer better management of persistent entity states.
> 
> ## Suggestions for additional features
> To complement your list of server setters, it might be interesting to add :
> 
> ### Setters for Dynamic Entities
> * SetPedComponentVariationServerSide**: To manage NPC clothing directly on the server side, which would improve scenarios involving uniforms, for example.
> * SetEntityVisibleServerSide**: To manage entity visibility without the need for a client call.
> 
> ### Improved synchronization
> One of the major benefits of server-side setters is improved entity synchronization, particularly in environments with a large number of simultaneous interactions (e.g. RP servers). If the server can handle more entity-related aspects, this would avoid frequent desynchronization, especially on OneSync servers with lots of players.
> 
> ### Performance optimization
> Reducing RPC calls could not only improve stability, but also lighten the load on the client side. By delegating more to the server, this would enable a better distribution of tasks, particularly beneficial for clients with more modest configurations.
> 
> ## Conclusion
> This proposal to add server-side setters would be a major step forward in improving the stability, security and maintainability of FiveM servers. I strongly encourage this initiative and am curious to see how it could be integrated with the framework's current features.
> 
> Thank you for your suggestion, which I hope will be taken on board by the CFX team.

Can you not post these meaningless chatGPT replies, they do nothing helpful and only block any actual discussion.

**Comment by Firou91** [comment]
> > # Server Setters for Authority and Stability (Response to Feature Request)
> > I find this request very relevant, as it addresses key issues that developers often encounter due to over-reliance on native RPC. Here are some thoughts and additions on the subject:
> > ## Current problems
> > ### Instability of native RPC
> > It's true that RPC natives present reliability problems, especially when it comes to server-side entity manipulation. This leads not only to random bugs, but also to inconsistencies in entity synchronization between clients. Many developers circumvent these limitations by using delays and other less-than-optimal methods. It would indeed be beneficial to have reliable server setters to get around these difficulties.
> > ### Lack of Server Authority
> > Server-side entity management is limited by the current implementation, which leaves a great deal of responsibility to the client. Introducing server-setters would not only improve stability, but also enhance entity security and integrity. This would prevent unwanted client-side manipulation and offer better management of persistent entity states.
> > ## Suggestions for additional features
> > To complement your list of server setters, it might be interesting to add :
> > ### Setters for Dynamic Entities
> > 
> > * SetPedComponentVariationServerSide**: To manage NPC clothing directly on the server side, which would improve scenarios involving uniforms, for example.
> > * SetEntityVisibleServerSide**: To manage entity visibility without the need for a client call.
> > 
> > ### Improved synchronization
> > One of the major benefits of server-side setters is improved entity synchronization, particularly in environments with a large number of simultaneous interactions (e.g. RP servers). If the server can handle more entity-related aspects, this would avoid frequent desynchronization, especially on OneSync servers with lots of players.
> > ### Performance optimization
> > Reducing RPC calls could not only improve stability, but also lighten the load on the client side. By delegating more to the server, this would enable a better distribution of tasks, particularly beneficial for clients with more modest configurations.
> > ## Conclusion
> > This proposal to add server-side setters would be a major step forward in improving the stability, security and maintainability of FiveM servers. I strongly encourage this initiative and am curious to see how it could be integrated with the framework's current features.
> > Thank you for your suggestion, which I hope will be taken on board by the CFX team.
> 
> Can you not post these meaningless chatGPT replies, they do nothing helpful and only block any actual discussion.

Okay if you say so. I don't know if it's Deepl's translation that gives that impression, but it's not. 😑
So please, if you're the one who censored my quote reply, please remove it.

**Comment by Firou91** [comment]
> > # Server Setters for Authority and Stability (Response to Feature Request)
> > I find this request very relevant, as it addresses key issues that developers often encounter due to over-reliance on native RPC. Here are some thoughts and additions on the subject:
> > ## Current problems
> > ### Instability of native RPC
> > It's true that RPC natives present reliability problems, especially when it comes to server-side entity manipulation. This leads not only to random bugs, but also to inconsistencies in entity synchronization between clients. Many developers circumvent these limitations by using delays and other less-than-optimal methods. It would indeed be beneficial to have reliable server setters to get around these difficulties.
> > ### Lack of Server Authority
> > Server-side entity management is limited by the current implementation, which leaves a great deal of responsibility to the client. Introducing server-setters would not only improve stability, but also enhance entity security and integrity. This would prevent unwanted client-side manipulation and offer better management of persistent entity states.
> > ## Suggestions for additional features
> > To complement your list of server setters, it might be interesting to add :
> > ### Setters for Dynamic Entities
> > 
> > * SetPedComponentVariationServerSide**: To manage NPC clothing directly on the server side, which would improve scenarios involving uniforms, for example.
> > * SetEntityVisibleServerSide**: To manage entity visibility without the need for a client call.
> > 
> > ### Improved synchronization
> > One of the major benefits of server-side setters is improved entity synchronization, particularly in environments with a large number of simultaneous interactions (e.g. RP servers). If the server can handle more entity-related aspects, this would avoid frequent desynchronization, especially on OneSync servers with lots of players.
> > ### Performance optimization
> > Reducing RPC calls could not only improve stability, but also lighten the load on the client side. By delegating more to the server, this would enable a better distribution of tasks, particularly beneficial for clients with more modest configurations.
> > ## Conclusion
> > This proposal to add server-side setters would be a major step forward in improving the stability, security and maintainability of FiveM servers. I strongly encourage this initiative and am curious to see how it could be integrated with the framework's current features.
> > Thank you for your suggestion, which I hope will be taken on board by the CFX team.
> 
> Can you not post these meaningless chatGPT replies, they do nothing helpful and only block any actual discussion.

I'm not going to bother arguing with you... I didn't know Deepl's translation made it sound so “**CHATGPT**” 🤷‍♂️. Honestly, stop making fabrications based on nonsense...

I've simply repeated what I understood and added my own words. If you don't like it, I can make a shorter reply.

**Comment by n3xuuu** [comment]
would love to see this happening

**Comment by alberttheprince** [comment]
We're currently experiencing issues with spawning vehicles server side, as it's not being correctly set when the vehicle spawned, it happens enough to where it's nearly a daily occurrence with a small player base of 10-20 a day, I would imagine this is much worse with larger number of players.

**Comment by BennoArnautovic** [comment]
> > 👍🏻 Good idea we need this !!!!
> 
> Well thank god you and the 30 other muppets posted "good idea!!!". Really contributing to the discussion there champ.

Well, how can we muppets help and contribute to this discussion. For example, had an idea and plan how to create complex cayo perico zombie region, with server side zombies. But due to huge lack of RPC natives for entity manipulation.

**Comment by Morji789** [comment]
You can do the effect of server setter "SetVehicleDoorsLocked" (for example) in the client by using state bags

**Comment by lLukDreams** [comment]
Absolutely agree with this feature request. The current reliance on RPC natives introduces significant instability and complexity to development. It's frustrating to implement unreliable workarounds for basic entity manipulations that should be handled server-side with confidence.

As someone migrating from other platforms, I'm honestly surprised that such a fundamental feature—direct server-side control over entities—is missing. It’s something I took for granted elsewhere, and its absence is really felt here. I’d love to see not just these setters implemented, but broader discussion and development around giving servers more authoritative control in general.

Server Setters would be a huge step forward for developer experience and code quality. They would provide:

Stability: Reducing flaky behavior from RPC calls.

Authority: Giving the server the final say over entity states.

Maintainability: Cleaner, more predictable code without spaghetti logic or delay hacks.

This change wouldn’t just benefit individual developers—it would lead to more reliable and secure community servers across the board. I strongly support prioritizing this.

**Comment by antond15** [comment]
Would be really great :)

**Comment by imsnaily** [comment]
is this feature request gonna stay opened till the platform dies or wtw

**Comment by mcNuggets1** [comment]
You aint the first one with this idea. I'd argue nobody would do the associated work and also that fivem currently is a huge bandaidfix.
